functions/Invoke-DbaLogShippingRecovery.ps1

```diff
@@ -0,0 +1,307 @@
+		}
+		else
+		{
+            # If just one database is set in the database parameter
+			if($Database -ne $null)
+            {
+			    $db = $Database[0]
+
+                $query += "WHERE lsd.secondary_database = '$db'"
+            }
+		}
+
+        # Retrieve the log shipping information from the secondary instance
+        try{
+            Write-Verbose "Retrieving log shipping information from the secondary instance"
+            $logshipping_details = Invoke-Sqlcmd2 -ServerInstance $SqlServer -Database 'msdb' -Query $query
+        }
+        catch
+        {
+            Write-Error $_.Exception.Message
+            Break
+        }
+        
+        # Check if there are any databases to recover
+        if($logshipping_details -ne $null)
+        {
+
+            # Loop through each of the log shipped databases
+            foreach($ls in $logshipping_details)
+            {
+                # Check if the database is in the right state
+                if($server.Databases[$ls.secondary_database].Status -in ('Normal, Standby', 'Standby', 'Restoring'))
+                {
+
+                    Write-Output ("Started Recovery for '" + $ls.secondary_database + "'")
+
+                    # Check if the backup source directory can be reached
+                    if(Test-Path $ls.backup_source_directory)
+                    {
+                        # Get the last file from the backup source directory
+                        $latestfile = Get-ChildItem -Path $ls.backup_source_directory -filter ("*" + $ls.primary_database + "*") | where {($_.Extension -eq '.trn') } | Sort-Object LastWriteTime -Descending | Select-Object -First 1
+                    
+                        # Check if the latest file is also the latest copied file
+                        if($latestfile.Name -ne [string]$ls.last_copied_file.Split('\')[-1])
+                        {
+                            Write-Verbose "Synchronizing the latest transaction log backup file(s)"
+
+                            # Start the job to get the latest files
+                            if($PSCmdlet.ShouldProcess($SqlServer, ("Starting copy job " + $ls.copyjob)))
+                            {
+                                $server.JobServer.Jobs[$ls.copyjob].Start()
+
+                                Write-Verbose ("Waiting for files to get copied to '" + $ls.backup_destination_directory + "'")
+
+                            }
+
+                            # Check if the file has been copied
+                            $query = "SELECT last_copied_file FROM msdb.dbo.log_shipping_secondary WHERE primary_database = '" + $ls.primary_database + "'"
+                            $latestcopy = Invoke-Sqlcmd2 -ServerInstance $SqlServer -Database 'msdb' -Query $query
+
+                            # Waiting for the copy action to complete
+                            if($PSCmdlet.ShouldProcess($ls.backup_destination_directory, "Waiting for the copy action to complete.. "))
```

Comment:
I can probably put more actions into one shouldprocess check. I'm going to look into that.

Normally log shipping is set up using sql server agent jobs. The jobs have either backup, copy or restore action. The copy and restore jobs are on the secondary instance. Those jobs use the sqllogship.exe which takes care of a lot of checks and internals.
Te reason I use the jobs to execute the copy and restore is because I'd have to make lots of extra code. One of the steps I take is disabling the copy and restore jobs after they've been executed which removes the possibility of errors.
If I'd have to do a disaster recovery with log shipping I'd do the same kind of actions but manually. In my production environment I have over 25 databases being log shipped which would probably take quite some time to restore.